### PR TITLE
DOMA-3744 Client card

### PR DESCRIPTION
**Pull request by nomerdvadcatpyat**
The client card is needed so that the employee can quickly find the necessary information about the client, his tickets, property hint and so on.

изображение
изображение

### CODE REVIEW

File: apps/condo/pages/phone/[number]/index.tsx [hunk 1]
```diff
@@ -0,0 +1,597 @@
+import { Col, Row, Tabs, Typography } from 'antd'
+import { EllipsisConfig } from 'antd/es/typography/Base'
+import { get, uniqBy } from 'lodash'
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:3]: same

File: apps/condo/pages/phone/[number]/index.tsx [hunk 2]
```diff
@@ -0,0 +1,597 @@
...
+    const handleRowAction = useCallback((record) => {
+        return {
+            onClick: async () => {
+                await router.push(`/ticket/${record.id}/`)
+            },
+        }
+    }, [router])
+
+    return (
+        <Row gutter={[0, 60]}>
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:216]: move to constant pls

File: apps/condo/pages/phone/[number]/index.tsx [hunk 3]
```diff
@@ -0,0 +1,597 @@
...
+            <Col span={24}>
+                <Row gutter={[0, 40]}>
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:218]: same

File: apps/condo/pages/phone/[number]/index.tsx [hunk 4]
```diff
@@ -0,0 +1,597 @@
...
+                    <Col span={24}>
+                        <TicketPropertyHintCard
+                            propertyId={property.id}
+                            hintContentStyle={HINT_CARD_STYLE}
+                        />
+                    </Col>
+                </Row>
+            </Col>
+            <Col span={24}>
+                <Row gutter={[0, 24]}>
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:236]: same

File: apps/condo/pages/phone/[number]/index.tsx [hunk 5]
```diff
@@ -0,0 +1,597 @@
...
+                            onRow={handleRowAction}
+                            data-cy='ticket__table'
+                        />
+                    </Col>
+                </Row>
+            </Col>
+            <ActionBar>
+                <Button
+                    key='submit'
+                    onClick={() => handleTicketCreateClick(lastCreatedTicket)}
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:257]: useCallback

File: apps/condo/pages/phone/[number]/index.tsx [hunk 6]
```diff
@@ -0,0 +1,597 @@
...
+  align-items: center;
+`
+
+const ADD_ADDRESS_TEXT_STYLE: CSSProperties = { marginTop: '12px' }
+
+const AddAddressTabTitle = () => {
+    const intl = useIntl()
+    const AddAddressMessage = intl.formatMessage({ id: 'pages.clientCard.addAddress' })
+
+    const [active, setActive] = useState<boolean>()
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:425]: Is it possible to do the same using ":hover" in css?

File: apps/condo/pages/phone/[number]/index.tsx [hunk 7]
```diff
@@ -0,0 +1,597 @@
...
+    }, [phoneNumber, router])
+
+    return (
+        <>
+            <Head>
+                <title>{ClientCardTitle}</title>
+            </Head>
+            <PageWrapper>
+                <PageContent>
+                    <Row gutter={[0, 60]}>
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:479]: move to const

File: apps/condo/pages/phone/[number]/index.tsx [hunk 8]
```diff
@@ -0,0 +1,597 @@
...
+                                                <ClientTabContent
+                                                    type={type}
+                                                    property={property}
+                                                    phone={phoneNumber}
+                                                    unitName={unitName}
+                                                    canManageContacts={canManageContacts}
+                                                />
+                                            </Tabs.TabPane>
+                                        )
+                                    })
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:519]: Perhaps you should memoize these elements?

File: apps/condo/pages/phone/[number]/index.tsx [hunk 9]
```diff
@@ -0,0 +1,597 @@
...
+
+    const { objs: employees } = OrganizationEmployee.useObjects({
+        where: {
+            ...baseQuery,
+            phone: phoneNumber,
+        },
+    })
+
+    const employeeTickets = tickets.filter(ticket => !!employees.find(employee => employee.phone === ticket.clientPhone))
+    const notResidentTickets = tickets.filter(ticket => !employeeTickets.find(employeeTicket => employeeTicket.id === ticket.id))
```
Comment (Alllex202) [inline apps/condo/pages/phone/[number]/index.tsx:557]: useMemo?

File: apps/condo/domains/common/utils/filters.utils.ts [hunk 10]
```diff
@@ -186,25 +186,26 @@ export function convertToOptions <T> (objects: T[], labelField: string, valueFie
...
     const possibleFilters = pickBy(newFilters, (value) => !isEmpty(value))
-    const possibleQueryData = { ...router.query, sort, offset }
+    const possibleQueryData = newRoute ? { sort, offset } : { ...router.query, sort, offset }
     if (isEmpty(possibleFilters)) {
         delete possibleQueryData['filters']
     } else {
         possibleQueryData['filters'] = JSON.stringify(possibleFilters)
     }
 
+    const route = newRoute ? newRoute : router.route
```
Comment (AntonAL) [inline apps/condo/domains/common/utils/filters.utils.ts:202]:
```suggestion
    const route = newRoute || router.route
```

File: apps/condo/domains/contact/components/ContactsEditor/useContactsEditorHook.tsx [hunk 11]
```diff
@@ -5,10 +5,9 @@ import { PROPERTY_REQUIRED_ERROR } from '@condo/domains/common/constants/errors'
 import { BuildingUnitSubType, Contact as ContactType, OrganizationEmployeeRole } from '@app/condo/schema'
 
 interface IContactsEditorHookArgs {
-    // Organization scope for contacts autocomplete and new contact, that can be created
-    organization: string,
     role?: OrganizationEmployeeRole,
     allowLandLine?: boolean,
+    initialQuery?: unknown
```
Comment (AntonAL) [inline apps/condo/domains/contact/components/ContactsEditor/useContactsEditorHook.tsx:10]: Is it possible to know type in advance, at least a set of not required fields?

File: apps/condo/domains/contact/utils/clientCard.tsx [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+        const { data, error } = await _search(client, SEARCH_BY_PHONE, { phone, organizationId })
+
+        if (error) console.warn(error)
+
+        const employees = []
+        const contacts = data.contacts.map(({ id, phone, property, unitName }) => ({
+            value: id, phone, property, unitName, type: ClientType.Resident,
+        }))
+        const tickets = data.tickets.filter(ticket => {
+            const employee = data.employees.find(employee => employee.phone === ticket.clientPhone)
```
Comment (AntonAL) [inline apps/condo/domains/contact/utils/clientCard.tsx:80]: Why you filtering result by phone when GraphQL query already has filters?

File: apps/condo/domains/contact/components/ContactsEditor/index.tsx [hunk 13]
```diff
@@ -118,14 +119,14 @@ export const ContactsEditor: React.FC<IContactEditorProps> = (props) => {
...
         property: { id: property ? property : null },
         unitName: unitName ? unitName : undefined,
         unitType: unitType ? unitType : undefined,
-    }), [organization, property, unitName, unitType])
+    }), [initialQuery, property, unitName, unitType])
 
     const initialEmployeesQuery = useMemo(() => ({
-        organization: { id: organization },
+        ...initialQuery,
     }), [organization])
```
Comment (SavelevMatthew) [inline apps/condo/domains/contact/components/ContactsEditor/index.tsx:130]: Why no initial query in deps?

File: apps/condo/domains/contact/utils/clientCard.tsx [hunk 14]
```diff
@@ -0,0 +1,145 @@
...
+                <Col>
+                    {property ? getAddressRender(property) : DeletedMessage}
+                </Col>
+            </Row>
+        </Typography.Link>
+    </Select.Option>
+)
+
+export const getClientCardTabKey = (propertyId: string, type: ClientType, unitName: string): string =>
+    `${propertyId}-${unitName ? unitName : ''}-${type}`
```
Comment (AntonAL) [inline apps/condo/domains/contact/utils/clientCard.tsx:122]:
```suggestion
    compact([propertyId, unitName, type]).join('-')
```

File: apps/condo/domains/contact/utils/clientCard.tsx [hunk 15]
```diff
@@ -0,0 +1,145 @@
...
+                </Col>
+                <Col>
+                    {property ? getAddressRender(property) : DeletedMessage}
+                </Col>
+            </Row>
+        </Typography.Link>
+    </Select.Option>
+)
+
+export const getClientCardTabKey = (propertyId: string, type: ClientType, unitName: string): string =>
```
Comment (AntonAL) [inline apps/condo/domains/contact/utils/clientCard.tsx:121]: If `unitName` is not required, you can mark it appropriately
```suggestion
export const getClientCardTabKey = (propertyId: string, type: ClientType, unitName: string?): string =>
```

**Review by SavelevMatthew — APPROVED** [review]

File: apps/condo/domains/contact/components/ContactsEditor/index.tsx (hunk 13 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/contact/components/ContactsEditor/index.tsx:130]: missed

File: apps/condo/domains/contact/utils/clientCard.tsx (hunk 12 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/contact/utils/clientCard.tsx:80]: we need to find tickets from employees => such not resident tickets, where `clientPhone` === `employee.phone`

**Review by Alllex202 — APPROVED** [review]

File: apps/condo/domains/common/components/PhoneInput.tsx [hunk 16]
```diff
@@ -8,21 +8,27 @@ import React, {
...
 
-interface IPhoneInputProps extends InputProps {
+interface IPhoneInputProps extends Omit<PhoneInputProps, 'onChange'> {
     block?: boolean
     autoFormat?: boolean
     /*
         Make this component compatible with `AutoComplete` component, when used as a custom input.
      */
     compatibilityWithAntAutoComplete?: boolean,
+    masks?: { ru?: string }
```
Comment (sitozzz) [inline apps/condo/domains/common/components/PhoneInput.tsx:26]: What about another languages?

File: apps/condo/domains/common/components/PhoneInput.tsx [hunk 17]
```diff
@@ -68,7 +74,7 @@ export const PhoneInput: React.FC<IPhoneInputProps> = forwardRef((props, ref) =>
     useEffect(() => {
         inputRef.current.numberInputRef.tabIndex = props.tabIndex
         // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (sitozzz) [inline apps/condo/domains/common/components/PhoneInput.tsx:76]: Seems like this comment is unnecessary now

File: apps/condo/domains/common/hooks/useSearchByPhoneModal.tsx [hunk 18]
```diff
@@ -0,0 +1,218 @@
...
+
+    const handleCloseModal = useCallback(() => setIsSearchByPhoneModalVisible(false), [])
+
+    const SearchByPhoneModal = useMemo(() => (
+        <StyledModal
+            visible={isSearchByPhoneModalVisible}
+            title={SearchByPhoneMessage}
+            onCancel={handleCloseModal}
+            footer={null}
+            width={1150}
```
[the remark this answers is not in the document]
Comment (nomerdvadcatpyat) [reply apps/condo/domains/common/hooks/useSearchByPhoneModal.tsx:217]: yes

**Pull request merged**